src/ec/eddsa.rs

```diff
@@ -80,42 +77,86 @@ impl<'a> Ed25519KeyPair {
-        if private_key.len() != 32 {
+        if private_key.len() != PRIVATE_KEY_LEN {
             return Err(error::Unspecified);
         }
-        if public_key.len() != 32 {
+        if public_key.len() != PUBLIC_KEY_LEN {
             return Err(error::Unspecified);
         }
-        let mut pair = Ed25519KeyPair { private_public: [0; 64] };
-        for i in 0..32 {
+        let mut pair = Ed25519KeyPair { private_public: [0; KEY_PAIR_LEN] };
+        for i in 0..SCALAR_LEN {
             pair.private_public[i] = private_key[i];
-            pair.private_public[32 + i] = public_key[i];
+            pair.private_public[SCALAR_LEN + i] = public_key[i];
         }
         Ok(pair)
     }
 
+    // Returns a reference to the little-endian-encoded private key bytes.
+    fn private_key_bytes(&'a self) -> &'a [u8] {
+        &self.private_public[..PRIVATE_KEY_LEN]
+    }
+
     /// Returns a reference to the little-endian-encoded public key bytes.
-    pub fn public_key_bytes(&'a self) -> &'a [u8] { &self.private_public[32..] }
+    pub fn public_key_bytes(&'a self) -> &'a [u8] {
+        &self.private_public[PRIVATE_KEY_LEN..]
+    }
 
     /// Returns the signature of the message `msg`.
     pub fn sign(&self, msg: &[u8]) -> signature::Signature {
-        let mut signature_bytes = [0u8; 64];
+        let mut signature_bytes = [0u8; SIGNATURE_LEN];
+        let mut az = [0u8; digest::SHA512_OUTPUT_LEN];
+        az.copy_from_slice(digest::digest(&digest::SHA512,
+                                          &self.private_key_bytes()).as_ref());
+        unsafe {
+            GFp_ed25519_scalar_mask(slice_as_scalar_ref_mut(&mut az).unwrap())
+        };
+        let mut nonce = [0u8; digest::SHA512_OUTPUT_LEN];
+        nonce.copy_from_slice({
+            let mut ctx = digest::Context::new(&digest::SHA512);
+            ctx.update(&az[SCALAR_LEN..]);
+            ctx.update(&msg);
+            ctx.finish().as_ref()
+        });
+        unsafe {
+            GFp_x25519_sc_reduce(slice_as_scalar_ref_mut(&mut nonce).unwrap())
+        };
+        let mut r = ExtPoint::new_at_infinity();
+        unsafe {
+            GFp_x25519_ge_scalarmult_base(&mut r, slice_as_scalar_ref_mut(&mut nonce).unwrap());
+            GFp_ge_p3_tobytes(slice_as_scalar_ref_mut(&mut signature_bytes).unwrap(), &r);
+        }
+        let mut hram = [0u8; digest::SHA512_OUTPUT_LEN];
+        hram.copy_from_slice({
+            let mut ctx = digest::Context::new(&digest::SHA512);
+            ctx.update(&signature_bytes[..SCALAR_LEN]);
+            ctx.update(&self.public_key_bytes());
+            ctx.update(&msg);
+            ctx.finish().as_ref()
+        });
         unsafe {
-            GFp_ed25519_sign(signature_bytes.as_mut_ptr(), msg.as_ptr(),
-                             msg.len(), self.private_public.as_ptr());
+            GFp_x25519_sc_reduce(slice_as_scalar_ref_mut(&mut hram).unwrap());
```

Comment:
This worked well! I feel like the name could be improved, but I stuck with the one you suggested to be consistent with the other code.